netann/host_ann.go

```diff
@@ -0,0 +1,198 @@
+	// changes.
+	RefreshTicker ticker.Ticker
+
+	// LookupHost performs DNS resolution on a given host and returns its
+	// addresses.
+	LookupHost func(string) (net.Addr, error)
+
+	// AdvertisedIPs is the set of IPs that we've already announced with
+	// our current NodeAnnouncement. This set will be constructed to avoid
+	// unnecessary node NodeAnnouncement updates.
+	AdvertisedIPs map[string]struct{}
+
+	// AnnounceNewIPs announces a new set of IP addresses for the backing
+	// Lightning node. The first set of addresses is the new set of
+	// addresses that we should advertise, while the other set are the
+	// stale addresses that we should no longer advertise.
+	AnnounceNewIPs func([]net.Addr, map[string]struct{}) error
+}
+
+// HostAnnouncer is a sub-system that allows a user to specify a set of hosts
+// for lnd that will be continually resolved to notice any IP address changes.
+// If the target IP address for a host changes, then we'll generate a new
+// NodeAnnouncement that includes these new IPs.
+type HostAnnouncer struct {
+	cfg HostAnnouncerConfig
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+
+	startOnce sync.Once
+	stopOnce  sync.Once
+}
+
+// NewHostAnnouncer returns a new instance of the HostAnnouncer.
+func NewHostAnnouncer(cfg HostAnnouncerConfig) *HostAnnouncer {
+	return &HostAnnouncer{
+		cfg:  cfg,
+		quit: make(chan struct{}),
+	}
+}
+
+// Start starts the HostAnnouncer.
+func (h *HostAnnouncer) Start() error {
+	h.startOnce.Do(func() {
+		h.wg.Add(1)
+		go h.hostWatcher()
+	})
+
+	return nil
+}
+
+// Stop signals the HostAnnouncer for a graceful stop.
+func (h *HostAnnouncer) Stop() error {
+	h.stopOnce.Do(func() {
+		close(h.quit)
+		h.wg.Wait()
+	})
+
+	return nil
+}
+
+// hostWatcher periodically attempts to resolve the IP for each host, updating
+// them if they change within the interval.
+func (h *HostAnnouncer) hostWatcher() {
+	defer h.wg.Done()
+
+	ipMapping := make(map[string]net.Addr)
+	refreshHosts := func() {
+
+		// We'll now run through each of our hosts to check if they had
+		// their backing IPs changed. If so, we'll want to re-announce
+		// them.
+		var addrsToUpdate []net.Addr
+		addrsToRemove := make(map[string]struct{})
+		for _, host := range h.cfg.Hosts {
+			newAddr, err := h.cfg.LookupHost(host)
+			if err != nil {
+				log.Warnf("unable to resolve IP for "+
+					"host %v: %v", host, err)
+				continue
+			}
+
+			// If nothing has changed since the last time we
+			// checked, then we don't need to do any updates.
+			oldAddr, oldAddrFound := ipMapping[host]
+			if oldAddrFound && oldAddr.String() == newAddr.String() {
+				continue
+			}
+
+			// Update the IP mapping now, as if this is the first
+			// time then we don't need to send a new announcement.
+			ipMapping[host] = newAddr
+
+			// If this IP has already been announced, then we'll
+			// skip it to avoid triggering an unnecessary node
+			// announcement update.
+			_, ipAnnounced := h.cfg.AdvertisedIPs[newAddr.String()]
```

Comment:
Nice! Not sure how likely it is for a host to resolve to a new address and go back to the previous address, but it may be worth updating this set after we announce to only contain the new addresses.